### ISSUE TITLE
20s WebGL inference for first pass of <3MB model on RTX 3090, crashes altogether on mobile

### ISSUE THREAD

**Issue by jphdotam** [tensorflow/tfjs#6970]
**System information**
- I have written a standard segmentation network and converted it to TFJS format successfully.
- Windows with Edge, Macbook pro on Safari, iPhone with Chrome. Desktop with RTX 3090 and latest NVidia drivers running Windows. Intel core i9 9900X.
- TensorFlow.js script link https://cdn.jsdelivr.net/npm/@tensorflow/tfjs@4.0.0/dist/tf.min.js but version 2.*  also tried. Version 1.* lacks operations preventing model inference.
- TensorFlow.js version (use command below): 4.0 and below (1.2 etc. misses some key operations)
- Browser version: Edge v106.0.1370.47 (other browsers also tried)

**Description of Problem**
I have a simple UNet network with EfficientNetB1 as a backbone.
It works correctly, and I get the correct proedictions in tensorflow-js.
However, running a simple `predict()` on a single 320 * 320 image freezes my high spec desktop for up to 20 seconds. However, further inferences using the same tensor size (but a different image) take 0.05s. Increasing the batch size causes another 20s delay (i.e. the warmup is specific to the batch size - maybe this is normal?).
I have tried lots of things to improve this
- A ResNet based model, rather than EfficientNet - almost no change (maybe slightly faster).
- Using `tf.ENV.set('WEBGL_PACK', false)` - this slightly improves the forward pass but it still crashes.
- Enabling Debug mode (see alternative link below for that example). In version 2.0.0 of tensorflow-js it gave me a warning message about using > 2Gb of GPU RAM and a memory leak, but by using version 4.0.0 this message disappears. I'm unsure if this reflects a fixed problem, or just a higher threshold for warning messages!

**Standalone code to reproduce the issue**
I have an example script at http://cardiologists.london/static/atria/atria_tfjs_uint8.htm - this uses a uint8 quantised UNet with a simple EfficientNetB1 backbone.
A version with Debug mode enabled for the predict() function is here: 
http://cardiologists.london/static/atria/old/atria.htm
The embedded JS file shows the model location

**Other info / logs**
The javascript console log for Debug mode is here: http://cardiologists.london/static/atria/cardiologists.london-1666617537754.log

**Comment by pyu10055** [comment]
@jphdotam Based on your description, I believe the initial delay is caused by the large memory footprint, I also looked at your debug page, I notice your model uses conv2d with larger filters.
I will suggest you to enable this flag 'WEBGL_EXP_CONV' to use a more optimized conv2d kernel implementation to reduce the inference time and memory footprint.
`tf.env().set('WEBGL_EXP_CONV', true)`
Please give it a try and let us know if that works for you.

**Comment by jphdotam** [comment]
Thanks so much for your input @pyu10055 , but when I tried that I actually found an **increase** in time of around 15% (averaged over several attempts...)

**Comment by pyu10055** [comment]
@jphdotam interesting, do you still have the initial 20s lock out?

**Comment by jphdotam** [comment]
Yes, sadly @pyu10055 . I'm not sure if it's similarly long for you when you access it? :
[This is the webpage with the displayed inference](http://cardiologists.london/static/atria/atria_tfjs_uint8.htm)
[And here is the JS file it includes with the added WEBGL_EXP_CONV flag](http://cardiologists.london/static/atria/atrai_tfjs_uint8.js)

**Comment by pyu10055** [comment]
@jphdotam To use the WEBGL_EXP_CONV flag, you cannot disable WEBGL_PACK flag
simply comment out this line `tf.ENV.set('WEBGL_PACK', false)`

**Comment by jphdotam** [comment]
Apologies. I've now done that and unfortunately the warmup timing is unchanged.

**Comment by jphdotam** [comment]
Ok there is something peculiar, I think, about WebGL inference using EfficientNets and MobileNets. I have now trained a ResNet18 based model (which has more parameters than mobilenetv3 and efficientnetb1) and it is more than 3 times the speed, despite being a larger weights file.

Efficientnet uses MobileNet blocks, of course, and so might it be the MBConv blocks with the inverted residuals or squeeze/excitation blocks that are causing the huge slowdown?

**Comment by pyu10055** [comment]
@jphdotam that is interested, do you have examples for both converted models? Running them in debug mode usually can reveal if there are any ops causing the performance degrade.

**Comment by Linchenn** [comment]
I saw you commented out `tf.env().set('WEBGL_EXP_CONV', true)`, instead of `tf.ENV.set('WEBGL_PACK', false)` in [And here is the JS file it includes with the added WEBGL_EXP_CONV flag](http://cardiologists.london/static/atria/atrai_tfjs_uint8.js)? I ran your codes on my MacBook, the first inference time drops from 6.195s to 4.879s with `WEBGL_EXP_CONV`. On debug mode, most of Conv ops' inference time drops, even though still time-consuming.
https://github.com/Linchenn/Toolkit/blob/main/debugTFJS/issue6970/index.js#L83

**Comment by Linchenn** [comment]
@jphdotam Could you try to add `tf.env().set('WEBGL_USE_SHAPES_UNIFORMS', true);` after the `tf.env().set('WEBGL_EXP_CONV', true);`? 

From my MacBook, the first inference time drops to 2.594s.

**Comment by jphdotam** [comment]
Interesting, on my Apple Silicon Macbook air, this reduces inference to 1.5s, which is the same as the ResNet model, so solves the issue.

However, on Windows with my RTX 3090 the above change (http://cardiologists.london/static/atria/old/atria.htm), only reduces time from 20s to 12s. It's still significantly worse than the larger ResNet model (https://github.com/tensorflow/tfjs/issues/6970) which is about 4s.

Could this be an issue with NVidia?

Both using Microsoft edge as a browser.

**Comment by pyu10055** [comment]
@jphdotam The first time inference involves shader compilation, the uniform_shape flag helps to reduce the number of unique shaders per model.
Typically, we would need to warm up the model first with some test data, so all the shaders are compiled and cached on the GPU. Our Intel partneralso noticed a huge performance gap between linux and windows for webgl warmup time, there are some speculations that the disk access is not efficient on windows that cause the performance problem. @gyagp

**Comment by Linchenn** [comment]
FYI, following Ping's comment, for the time of first inference,

http://cardiologists.london/static/atria/old/atria.htm has compiled 186 shader programs;
http://cardiologists.london/static/atria/atria.htm compiled 86 shader programs;
This may be one factor causing the slowness of your model, even though I am not sure the average shader program sizes for the two models. You could see this by calling immediately `Object.keys(tf.backend().binaryCache)` after the first inference complete.

**Comment by jphdotam** [comment]
Interesting, for the slow model I get 184 shaders and for the one that is 3x faster I get 120 - so this doesn't seem to be linear. And the slow one is 10x slower on Windows/NVidia than my Apple Silicon.

**Comment by gyagp** [comment]
When you run a WebGL program in browser for the first time on a machine, shaders will be compiled and cached into disk by browser (Actually graphics driver will also do another compilation and shader caching, but it's not a topic here). This is fully done by CPU, no matter how powerful your GPU is. 
TFJS hasn't implemented the support of parallel shader compilation, which means your shaders are compiled in sequence, with one CPU core.
For shader compilation, actually it includes 2 phases, shader conversion and compilation, done in ANGLE project. On Windows, WebGL shader (GLSL) will be converted to HLSL first, then compiled by FXC. On macOS (I'm not familiar with this path), shader will be converted to MSL, then compiled by MSL shader compiler. 

If it's not the first time you run the same WebGL program, browser will load the shader cache from disk, which usually means much faster than the time with shader compilation.
However, recent Chrome M106 (I believe it's same for Edge as it follows Chrome version quite closely) has a regression and shader cache is totally disabled unfortunately. I reported the issue at https://bugs.chromium.org/p/chromium/issues/detail?id=1375204, and it got fixed right away. But we have to wait for another release (M107) to have shader cache again. Of course, you may try Chrome Canary, which already included the fix. 
Even if we have the shader cache, according to our local tests, Windows is much slower than Linux. We're still in discussion to see if it's a bug or not. 

Anyway, more investigation to understand the breakdown data of so many seconds is a good next step, which will help us to know if we can do further optimizations (We may not help if FXC compiler is slow, which is not new). I will ask some team member to do this in the following days, and get back to you.

**Comment by Linchenn** [comment]
Thank you Yang for the detailed explanation!

**Comment by Linchenn** [comment]
@gyagp Hi Yang, actually Na previously implemented WebGL's parallel compilation. We are considerring to add the related flag to API doc soon. Otherwise, users have no ideas about the power of the flags, except directly reaching out to us. 

@jphdotam For all the following changes, you could refer to https://github.com/Linchenn/Toolkit/commit/af7cff31cd56d7e5ba633a598dd4303425336788 .

For your case, compilation is time-consuming. Could you try to add the following codes before you run the first inference? The following codes changes compilation mode from sequence to parallel.
```js
    tf.env().set('ENGINE_COMPILE_ONLY', true);
    model.predict(inputTensor).dataSync();
    tf.backend().checkCompileCompletion();
    tf.backend().getUniformLocations();
    tf.env().set('ENGINE_COMPILE_ONLY', false);
```
On my Windows machie with a 3060 card, the first inference time drops from 6.985s to 1.452s (compilation takes 1.163s and first run takes 0.289s) and the second inference time is 0.058s.

By the way, we could measure the time of the program more precisely, by adding `outputTensor.dataSync();`, which confirms GPU rendering pipeline completes.

**Comment by jphdotam** [comment]
That is MUCH quicker, it's gone from 12s inference time for me down to 1.4s. Thank you, @Linchenn and @gyagp - would there be any issue with you making this the default behaviour for tfjs?

**Comment by pyu10055** [comment]
@jphdotam It great to know that these flags help to improve performance in your use cases. 
We will make some of the flags the default in the future releases, but some of the flags are still in experimental state, for example 'ENGINE_COMPILE_ONLY' flag, it has some limitations, we are planning to provide better documentation for them in the API website.

**Comment by vladmandic** [comment]
@jphdotam i love compile pass feature (been using it for few months), but best to wrap it in try...catch block as it does fail for some models and it doesn't support async operations (so if model requires async execution, compile pass will fail)
